File: scripts/cmpi_lookup_test.rb
# Usage: bundle exec ruby scripts/cmpi_lookup_test.rb
# See https://github.com/jaechow/cardinal/blob/master/centinel/lookup/cmpi00.php for PHP example
require './environment'

# Cardinal Sandbox credentials:
api_key = ENV.fetch('API_KEY')
api_id = ENV.fetch('API_IDENTIFIER')
org_unit = ENV.fetch('ORG_UNIT_ID')

# Generate the request signature as documented under:
# https://cardinaldocs.atlassian.net/wiki/spaces/CCen/pages/1619492942/Cardinal+cmpi+Messages#Generating-a-Signature-Value
timestamp_buffer = 10000
timestamp = (Time.now.to_f * 1000).to_i - timestamp_buffer
request_signature = Base64.strict_encode64(
  Digest::SHA256.digest("#{timestamp}#{api_key}")
).strip

# Wealthsimple will provide TabaPay an AccountId, from which TabaPay can extract
# the below card details:
card_number = "4000000000001091"
card_expiry_month = "02"
card_expiry_year = "2024"
card_currency_code = "840"

# Additional details that Wealthsimple will provide to new TabaPay API endpoint:
df_reference_id = "c17dea31-9cf6-0c1b8f2d3c5"
transaction_id = "ws_transaction-0001"
transaction_amount = "12345"
billing_address = {
  street1: "4 Jersey St",
  street2: "",
  city: "Boston",
  postal_code: "02215",
  state: "MA",
  country_code: "840",
}
billing_person = {
  first: "Pedro",
  last: "Martinez",
  middle: "",
  email: "cardinal.mobile.test@example.com",
}

cmpi_lookup = <<-XML
<CardinalMPI>
    <Algorithm>SHA-256</Algorithm>
    <Amount>#{transaction_amount}</Amount>
    <BillingAddress1>#{billing_address[:street1]}</BillingAddress1>
    <BillingAddress2>#{billing_address[:street2]}</BillingAddress2>
    <BillingCity>#{billing_address[:city]}</BillingCity>
    <BillingPostalCode>#{billing_address[:postal_code]}</BillingPostalCode>
    <BillingState>#{billing_address[:state]}</BillingState>
    <BillingCountryCode>#{billing_address[:country_code]}</BillingCountryCode>
    <BillingFirstName>#{billing_person[:first]}</BillingFirstName>
    <BillingMiddleName>#{billing_person[:middle]}</BillingMiddleName>
    <BillingLastName>#{billing_person[:last]}</BillingLastName>
    <BrowserColorDepth>32</BrowserColorDepth>
    <BrowserHeader>text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8</BrowserHeader>
    <BrowserJavaEnabled>true</BrowserJavaEnabled>
    <BrowserLanguage>en-CA</BrowserLanguage>
    <BrowserScreenHeight>980</BrowserScreenHeight>
    <BrowserScreenWidth>1080</BrowserScreenWidth>
    <BrowserTimeZone>200</BrowserTimeZone>
    <CardExpMonth>#{card_expiry_month}</CardExpMonth>
    <CardExpYear>#{card_expiry_year}</CardExpYear>
    <CardNumber>#{card_number}</CardNumber>
    <CurrencyCode>#{card_currency_code}</CurrencyCode>
    <DFReferenceId>#{df_reference_id}</DFReferenceId>
    <DeviceChannel>browser</DeviceChannel>
    <Email>#{billing_person[:email]}</Email>
    <IPAddress>67.17.219.20</IPAddress>
    <Identifier>#{api_id}</Identifier>
    <MsgType>cmpi_lookup</MsgType>
    <OrderNumber>#{transaction_id}</OrderNumber>
    <OrgUnit>#{org_unit}</OrgUnit>
    <Signature>#{request_signature}</Signature>
    <Timestamp>#{timestamp}</Timestamp>
    <TransactionType>C</TransactionType>
    <UserAgent>Mozilla/5.0 (Windows NT 6.1; WOW64; rv:30.0) Gecko/20100101 Firefox/30.0</UserAgent>
    <Version>1.7</Version>
</CardinalMPI>
XML

puts cmpi_lookup

ch = Curl::Easy.new("https://centineltest.cardinalcommerce.com/maps/txns.asp")
ch.verbose = true
ch.timeout = 10
ch.ssl_verify_host = false
ch.http_post(Curl::PostField.file('cmpi_msg', cmpi_lookup))

# TabaPay will proxy this XML response as-is back to Wealthsimple:
response_body = ch.body_str
puts response_body

File: go-pkg/cmpilookup/cmpilookup_test.go
package cmpilookup

import (
	"io/ioutil"
	"testing"
)

func TestGenerateRequestSignatureWithValidInputs(t *testing.T) {
	// Sample values from https://cardinaldocs.atlassian.net/wiki/spaces/CCen/pages/1619492942/Cardinal+cmpi+Messages#Generating-a-Signature-Value
	apiKey := "13f1fd1b-ab2d-4c1f-8e2c-ca61878f2a44"
	timestamp := "1485534293321"
	signature, err := GenerateRequestSignature(apiKey, timestamp)
	expectedSignature := "X5TupwjjpO9hg5qIHG2h9aMCekWiqbYkzPkXkPopFMw="
	if signature != expectedSignature || err != nil {
		t.Fatalf(`GenerateRequestSignature => %q, %v, want match for %q, nil`, signature, err, expectedSignature)
	}
}

func TestGenerateRequestSignatureWithInvalidApiKey(t *testing.T) {
	apiKey := ""
	timestamp := "1485534293321"
	_, err := GenerateRequestSignature(apiKey, timestamp)
	if err == nil {
		t.Fatal("GenerateRequestSignature should return error for empty apiKey")
	}
}

func TestGenerateRequestSignatureWithInvalidTimestamp(t *testing.T) {
	apiKey := "13f1fd1b-ab2d-4c1f-8e2c-ca61878f2a44"
	timestamp := "1609890356"
	_, err := GenerateRequestSignature(apiKey, timestamp)
	if err == nil {
		t.Fatal("GenerateRequestSignature should return error for invalid timestamp")
	}
}

func TestGetRequestTimestampWithBuffer(t *testing.T) {
	timestamp := GetRequestTimestampWithBuffer()
	if len(timestamp) != 13 {
		t.Fatal("Expected unix epoch timestamp in milliseconds (13 digits)")
	}
}

func TestGenerateRequestBodyXmlWithValidParams(t *testing.T) {
	params := RequestBodyParams{
		ApiId:                        "api-id-123",
		OrgUnit:                      "org-unit-123",
		RequestSignature:             "X5TupwjjpO9hg5qIHG2h9aMCekWiqbYkzPkXkPopFMw=",
		RequestTimestamp:             "1485534293321",
		CardNumber:                   "4000000000001091",
		CardExpiryMonth:              "02",
		CardExpiryYear:               "2024",
		OrderAmount:                  "12345",
		OrderAuthenticationIndicator: "01",
		OrderCurrencyCode:            "840",
		OrderNumber:                  "ws_transaction-0001",
		OrderProductCode:             "ACF",
		OrderTransactionMode:         "P",
		OrderTransactionType:         "C",
		BrowserColorDepth:            "32",
		BrowserHeader:                "text/html,application/xhtml+xml,application/xml;q=0.9,",
		BrowserJavaEnabled:           "true",
		BrowserJavascriptEnabled:     "true",
		BrowserLanguage:              "en-CA",
		BrowserScreenHeight:          "980",
		BrowserScreenWidth:           "1080",
		BrowserTimeZone:              "200",
		DeviceChannel:                "browser",
		DeviceReferenceId:            "c17dea31-9cf6-0c1b8f2d3c5",
		IpAddress:                    "67.17.219.20",
		UserAgent:                    "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:30.0) Gecko/20100101 Firefox/30.0",
		BillingAddressStreet1:        "4 Jersey St",
		BillingAddressStreet2:        "Unit 123",
		BillingAddressCity:           "Boston",
		BillingAddressPostalCode:     "02215",
		BillingAddressState:          "MA",
		BillingAddressCountryCode:    "840",
		BillingFirstName:             "Pedro",
		BillingLastName:              "Martinez",
		BillingMiddleName:            "Jaime",
		Email:                        "cardinal.mobile.test@example.com",
		MobilePhone:                  "+16175551234",
	}
	requestBody, err := GenerateRequestBodyXml(params)
	expectedRequestBody, _ := ioutil.ReadFile("test-fixtures/cmpi_request_output.xml")
	if requestBody != string(expectedRequestBody) || err != nil {
		t.Fatalf(`GenerateRequestBodyXml => %q, %v, want match for %q, nil`, requestBody, err, expectedRequestBody)
	}
}
